osier/technology.py

```diff
@@ -0,0 +1,258 @@
+
+    Instead, try the more explicit approach:
+
+    >>> my_unit = "10 MW**-1"
+    >>> my_unit = unyt_quantity.from_string(my_unit)
+    unyt_quantity(10., '1/MW')
+
+    However, inverse MWh cannot be converted from a string.
+    """
+
+    def __init__(self,
+                 technology_name,
+                 technology_type='base',
+                 capital_cost=0.0,
+                 om_cost_fixed=0.0,
+                 om_cost_variable=0.0,
+                 fuel_cost=0.0,
+                 capacity=0.0,
+                 default_power_units=MW,
+                 default_time_units=hr,
+                 default_energy_units=None) -> None:
+
+        self.technology_name = technology_name
+        self.technology_type = technology_type
+
+        self.unit_power = default_power_units
+        self.unit_time = default_time_units
+        self.unit_energy = default_energy_units
+
+        self.capacity = capacity
+        self.capital_cost = capital_cost
+        self.om_cost_fixed = om_cost_fixed
+        self.om_cost_variable = om_cost_variable
+        self.fuel_cost = fuel_cost
+
+    @property
+    def unit_power(self):
+        return self._unit_power
+
+    @unit_power.setter
+    def unit_power(self, value):
+        self._unit_power = _validate_unit(value, dimension="power")
+
+    @property
+    def unit_time(self):
+        return self._unit_time
+
+    @unit_time.setter
+    def unit_time(self, value):
+        self._unit_time = _validate_unit(value, dimension="time")
+
+    @property
+    def unit_energy(self):
+        return self._unit_energy
+
+    @unit_energy.setter
+    def unit_energy(self, value):
+        if value:
+            self._unit_energy = _validate_unit(value, dimension="energy")
+        else:
+            self._unit_energy = self._unit_power * self._unit_time
+
+    @property
+    def capacity(self):
+        return self._capacity
+
+    @capacity.setter
+    def capacity(self, value):
+        self._capacity = _validate_quantity(value, dimension="power")
+
+    @property
+    def capital_cost(self):
+        return self._capital_cost
+
+    @capital_cost.setter
+    def capital_cost(self, value):
+        self._capital_cost = _validate_quantity(value, dimension="spec_power")
+
+    @property
+    def om_cost_fixed(self):
+        return self._om_cost_fixed
+
+    @om_cost_fixed.setter
+    def om_cost_fixed(self, value):
+        self._om_cost_fixed = _validate_quantity(value, dimension="spec_power")
+
+    @property
+    def om_cost_variable(self):
+        return self._om_cost_variable
+
+    @om_cost_variable.setter
+    def om_cost_variable(self, value):
+        self._om_cost_variable = _validate_quantity(
+            value, dimension="spec_energy")
+
+    @property
+    def fuel_cost(self):
+        return self._fuel_cost
+
+    @fuel_cost.setter
+    def fuel_cost(self, value):
+        self._fuel_cost = _validate_quantity(value, dimension="spec_energy")
```

Comment:
I'm not very familiar with wrappers, but could this be re-written by looping through the attributes?

Reply:
You might be able to do that. I know it's possible to loop through class attributes once they've been instantiated (so, not helpful here). However, that isn't desirable behavior because the `@property` decorator makes it possible to propagate state changes. For example:

```py
class Person():
  def __init__(self, first, last):
    self.first = first
    self.last = last

  @property
  def full(self):
    return self.first + " " + self.last
```
Without `@property` you could change the last name attribute, but the full name wouldn't change. With `@property` the full name is changed automatically. Also, you can access `full` as if it were another attribute (essentially without ()). 

```py
albert = Person("Albert", "Einstein")
print(albert.last)  # Einstein
print(albert.full)  # Albert Einstein
albert.last = Frankenstein
print(albert.full)  # Albert Frankenstein
```

This is a helpful [introduction to setters, getters, and property.](https://www.youtube.com/watch?v=jCzT9XFZ5bw).